<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Create A Package • CreateAPackage</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Create A Package">
<meta property="og:description" content="CreateAPackage">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CreateAPackage</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/CreateAPackage.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Kayla-Morrell/CreateAPackage">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="CreateAPackage_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Create A Package</h1>
                        <h4 class="author">Kayla Interdonato</h4>
            <address class="author_afil">
      Roswell Park Comprehensive Cancer Center, Buffalo, NY<br><h4 class="author">Lori Shepherd</h4>
            <address class="author_afil">
      Roswell Park Comprehensive Cancer Center, Buffalo, NY<br><h4 class="date">2020-07-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/Kayla-Morrell/CreateAPackage/blob/master/vignettes/CreateAPackage.Rmd"><code>vignettes/CreateAPackage.Rmd</code></a></small>
      <div class="hidden name"><code>CreateAPackage.Rmd</code></div>

    </address>
</address>
</div>

    
    
<div id="create-a-package" class="section level1">
<h1 class="hasAnchor">
<a href="#create-a-package" class="anchor"></a>Create A Package</h1>
<p>A package is a collection of functions that work together in a cohesive manner to achieve a goal. It (should) include detailed documentation through man pages and vignettes and (should) be tested for accuracy and efficiency. <a href="https://cran.r-project.org/doc/manuals/r-release/R-exts.html">Writing R extensions</a> is a very detailed manual provided by CRAN that explains writing packages and what the structure of a package should look like. This vignette is going to walk you through making your first package in RStudio as well as how to submitt a package to <em>Bioconductor</em>.</p>
<div id="using-devtools-to-create-a-package" class="section level2">
<h2 class="hasAnchor">
<a href="#using-devtools-to-create-a-package" class="anchor"></a>Using <code>devtools</code> to create a package</h2>
<p>The <code>devtools</code> package provides a lot of options and utility for helping to construct a new package. You can get a list of all available <code>devtools</code> functions with <code><a href="https://rdrr.io/r/base/ls.html">ls("package:devtools")</a></code>.</p>
<p>Some useful references for using <code>devtools</code> to build packages are <a href="https://www.rstudio.com/wp-content/uploads/2015/03/devtools-cheatsheet.pdf">Rstudio Devtools Cheetsheet</a> and <a href="http://stat545.com/packages06_foofactors-package.html">Jennifer Bryan class</a>.</p>
<div id="package-shell" class="section level3">
<h3 class="hasAnchor">
<a href="#package-shell" class="anchor"></a>Package shell</h3>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">devtools</span>)
<span class="fu"><a href="https://devtools.r-lib.org//reference/create.html">create</a></span>(<span class="st">"myFirstPackage"</span>)</pre></body></html></div>
<p><code><a href="https://devtools.r-lib.org//reference/create.html">create()</a></code> will <em>create</em> all the necessary files and sub-directories that are required by R to be a valid package: DESCRIPTION, NAMESPACE, and R directory. The DESCRIPTION file contains the basic information about the package and you will have to edit so that the information is pertinent to your package. The NAMESPACE file describes which functions will be imports and exports of the namespace. The R directory will contain the R code files for your package.</p>
<p>After running <code><a href="https://devtools.r-lib.org//reference/create.html">create()</a></code>, the package has a vaild package structure which means it can be installed and loaded:</p>
<ul>
<li>
<code><a href="https://rdrr.io/pkg/myFirstPackage/man">library("myFirstPackage")</a></code>,</li>
<li>
<code><a href="https://rdrr.io/pkg/myFirstPackage/man">library(help="myFirstPackage")</a></code>.</li>
</ul>
</div>
<div id="version-control" class="section level3">
<h3 class="hasAnchor">
<a href="#version-control" class="anchor"></a>Version control</h3>
<p>It is an excellent idea to version control whenever creating a package and especially when collaborating on a project, where multiple users are allowed to make changes. Version control allows for a constant recored of changes that can be advanced or reverted if necessary.</p>
<p>Only a project can be version controlled and to make a directory a project in RStudio go to: <code>File -&gt; New Project</code>. In this case we started creating the directory so we will follow the prompts for the option <code>Use Existing Directory</code>. Now that it is a project we can go to <code>Tools -&gt; Version Control -&gt; Project Setup</code> and change the <code>Version Control System</code> to <code>Git</code>, then follow the prompts. Notice in the RStudio pane for environments/history/build there is a new tab named ‘Git’. The package can now start using <code>git</code> version control through making commits. To make a commit, you can go to the Git tab, select the check box next to any files that have been modified, added, or deleted that you would like to track, and select <code>commit</code>. Enter a new commit message in the window that pops up and select <code>commit</code>.</p>
<p>It is important to tell Git just who we are. In RStudio, select <code>Tools -&gt; Shell</code> and type the following making sure to substitue <em>your</em> user.email and user.name. If you have github we recommend using your email and user.name associated with github here.</p>
<pre><code>git config --global user.email "&lt;someemail@gmail.com&gt;"
git config --global user.name "&lt;githubUserName&gt;"</code></pre>
<p>We could stop here but we also would like to put the package on GitHub. These next steps assume you have a github account. First, in RStudio go to <code>Tools -&gt;  Global Options</code> and select Git/SVN. Ensure the paths are correct. If you have not linked an RStudio project to GitHub, select <code>Create RSA key</code>. Close the window. Click on <code>View public key</code> and copy the displayed public key. Now in a web browser, open your GitHub accout. Go to <code>Settings</code> and <code>SSH and GPG keys</code>. Click on the option for <code>New SSH key</code> and paste the public key that was copied. Also on GitHub, create a new repository with the same name as the one you created with <code><a href="https://devtools.r-lib.org//reference/create.html">create()</a></code> in RStudio. Back in RStudio, select <code>Tools -&gt; Shell</code> and type the following making usre to substitue your GitHub user.name and the new package name.</p>
<pre><code>git remote add origin https://github.com/&lt;github user.name&gt;/&lt;package repo name&gt;.git
git remote set-url origin git@github.com:&lt;github user.name&gt;/&lt;package repo name&gt;.git
git pull origin master
git push -u origin master</code></pre>
<p>For instance, this is what the commands would look like for me:</p>
<pre><code>git remote add origin https://github.com/Kayla-Morrell/myFirstPackage.git
git remote set-url origin git@github.com:Kayla-Morrell/myFirstPackage.git
git pull origin master
git push -u origin master</code></pre>
<p>The <code>git remote add</code> command will create a new connection to the remote repository url and assign it the shortname ‘origin’ for easy referencing moving forward. Then the <code>git remote set-url</code> command is going to switch the url of the remote from ‘https’, which is public read only access, to ‘SSH’ so that you as a developer can have read and write access to the repository. <code>git  pull</code> is going to fetch and download content from the remote repository and update the local repository to match. <code>git push</code> does the exact opposite, it will upload the local repo changes to the remote repo.</p>
<p>Now if you look in the RStudio tab for Git, the push and pull options are available. You can now push and pull from/to the local and GitHub repository version of your package.</p>
</div>
</div>
<div id="description-file" class="section level2">
<h2 class="hasAnchor">
<a href="#description-file" class="anchor"></a>DESCRIPTION file</h2>
<p><code>devtools</code> provides built in functions for building, checking and installing a package. The package we created earlier using <code><a href="https://devtools.r-lib.org//reference/create.html">create()</a></code> has a valid package structure but if we did <code><a href="https://devtools.r-lib.org//reference/check.html">check()</a></code> we will find the DESCRIPTION file needs to be updated. The information in the DESCRIPTION file needs to be reflective of your package. These are the fields that should be changed:</p>
<ul>
<li>Title: Should be brief, but descriptive.</li>
<li>Version: Should be 0.99.0 for pre-release packages.</li>
<li>
<a href="mailto:Authors@R" class="email">Authors@R</a>: Only one person should be listed as maintainer (<code>cre</code>). We do accept Author/Maintainer for this field, either can be used but not both.</li>
<li>Description: Should be relatively short but a detailed overview of what the package functionality entails.</li>
<li>License: Preferably a standard license, Bioconductor core packages typically use Artistic-2.0. If using a non-standard license, must include a LICENSE file in your package and use <code>file LICENSE</code> in this field.</li>
<li>LazyData: Should be set to false, we find that this field being true can slow down the loading of a package especially if it contains large data.</li>
</ul>
<p>Throughout the development of your package you may have to update the DESCRIPTION file for appropriate Depends, Imports, and Suggests fields as you incorporate more functionality from other packages. As the package is developing we also require having a <code>biocViews:</code> field in the DESCRIPTION file. This field will contain at least two biocViews categories that reflect the nature of the package.</p>
</div>
<div id="r-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#r-functions" class="anchor"></a>R functions</h2>
<p>Now we want to starting writing R functions. In RStudio you can open an empty file by doing <code>File -&gt; New File</code> and selecting <code>R Script</code>. Save the file in the R directory. Write your functions and document. You can either document functions manually or if you use roxygen you can use the devetools function <code><a href="https://devtools.r-lib.org//reference/document.html">document()</a></code>. See the <a href="https://cran.r-project.org/doc/manuals/r-release/R-exts.html">Writing R extensions</a> for manual creation of Rd files, which belong in the man directory, but roxygen is growing increasingly popular. Some helpful links for roxygen tags can be found at <a href="https://www.rstudio.com/wp-content/uploads/2015/03/devtools-cheatsheet.pdf">RStudio Devtools Cheatsheet</a> and <a href="https://cran.rstudio.com/web/packages/roxygen2/vignettes/rd.html">Roxygen Help</a>.</p>
<p>Some useful <code>devtools</code> commands while creating functions are:</p>
<ul>
<li>
<code><a href="https://devtools.r-lib.org//reference/load_all.html">load_all()</a></code> which loads all package functions in environment to test,</li>
<li>
<code><a href="https://devtools.r-lib.org//reference/check.html">check()</a></code> which checks the package (R CMD check),</li>
<li>
<code><a href="https://devtools.r-lib.org//reference/document.html">document()</a></code> which generates or updates any documentation files.</li>
</ul>
<p>Using the RStudio options <code>Build -&gt; Build and Reload</code> and <code>Build -&gt; Clean  and Rebuild</code> will also help with function creation.</p>
<p>It is also recommended to have a man page for you package. <code>devtools</code> provides a framework for this. To create the file that needs to be modified use the function <code>use_package_doc()</code>.</p>
<p>If you import any functions in your code, don’t forget to update the DESCRIPTION file for Depends, Imports, or Suggests. If the function provides essential functionality for users of your package, it belongs in Depends. It is unusual for more than three packages to be listed as Depends. For packages that provide functions, methods, or classes that are used inside your package namespace, they belong in Imports. Most packages will be listed here. For packages that are used in vignettes, examples, or conditional code, they should be listed as Suggests. This includes examples that may use annotation and/or experiment packages.</p>
</div>
<div id="testing" class="section level2">
<h2 class="hasAnchor">
<a href="#testing" class="anchor"></a>Testing</h2>
<p>It is highly recommened to add unit tests to your package. Unit tests unsure that the package is working as expected. The two main ways to test are using <code>RUnit</code> or <code>testthat</code>. <code>testthat</code> functionality is included in <code>devtools</code> by using <code>use_testthat()</code>. This function will set up the needed directory structure and add the package suggestion to the DESCRIPTION file. Here are some examples of the structure of tests for <code>testthat</code>:</p>
<ul>
<li>
<code>expect_identical()</code>,</li>
<li>
<code>expect_true()</code>,</li>
<li>
<code>expect_error()</code>.</li>
</ul>
<p>There are other options as well that are discussed in <a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf">testthat Wickham</a> and <a href="http://r-pkgs.had.co.nz/tests.html">testthat</a>.</p>
</div>
<div id="vignettes" class="section level2">
<h2 class="hasAnchor">
<a href="#vignettes" class="anchor"></a>Vignettes</h2>
<p>Vignettes are another major documentation piece to a package. More and more repository systems (CRAN, Bioconductor, ROpenSci) are making vignettes a standard requirement. Vignettes are contain a more indepth description and examples of the package usage. <code>devtools</code> also provides the function <code>use_vignette()</code> to set up the directory structure and the initial file for a vignette. For Bioconductor submissions we recommend changing the <code>output:</code> section for the vignette header to the following, which would require adding <code>BiocStyle</code> to the Suggests field in the DESCRIPTION file,</p>
<pre><code>output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2</code></pre>
<p>Or if <code>BiocStyle</code> is already installed on your system, you can also use RStudio to set up the vignette by doing <code>New File -&gt; Rmarkdown -&gt; From Template -&gt;  Bioconductor HTML/PDF Vignette</code>.</p>
<p>A helpful rmarkdown link, which is commonly used for vignette creation, can be found here: <a href="http://www.rstudio.com/wp-content/uploads/2016/03/rmarkdown-cheatsheet-2.0.pdf">rmarkdown cheatsheet</a>.</p>
<div id="bioconductor-standards" class="section level3">
<h3 class="hasAnchor">
<a href="#bioconductor-standards" class="anchor"></a>Bioconductor Standards</h3>
<p>Now that we have gone over the basics of how to create a package, we will review what we look for (generally) in Bioconductor packages. Being mindful of these guidelines while developing your package will help the whole submission process.</p>
<ol style="list-style-type: decimal">
<li>Proper coding and efficient coding:</li>
</ol>
<ul>
<li><a href="http://bioconductor.org/developers/how-to/efficient-code/">Efficient and Robust Code</a></li>
<li><a href="http://bioconductor.org/developers/how-to/web-query/">Query Web Resource</a></li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>Bioconductor interconnectivity and S4 classes</li>
</ol>
<ul>
<li>S4 over S3 classes</li>
<li>Reuse existing infrastructure first!
<ul>
<li>DNA/RNA - <code>Biostrings</code> <code>DNAstringset</code>
</li>
<li>Gene sets - <code>GSEABase</code> <code>GeneSet</code>
</li>
<li>Genomic intervals - <code>GenomicRanges</code> <code>Granges</code>
</li>
<li>Rectangular feature x sample data - (RNAseq count matrix, microarray) - <code>SummarizedExperiment</code>/<code>MuliAssayExperiment</code>
</li>
<li>Single cell - <code>SingleCellExperiment</code>
</li>
<li>Mass spec - <code>MSnbase</code>
</li>
<li>import/loading data
<ul>
<li>GTF, GFF, BED, BigWig, … <code>rtracklayer::import()</code>
</li>
<li>FASTA <code>Biostrings::readDNAStringSet()</code>
</li>
<li>SAM/BAM <code>Rsamtools::scanBam()</code>, <code>GenomicAlignments::readGAlignment*()</code>
</li>
<li>VCF <code>VariantAnnotation::readVcf()</code>
</li>
<li>FASTQ <code>ShortRead::readFastq()</code>
</li>
</ul>
</li>
</ul>
</li>
<li><a href="http://bioconductor.org/packages/release/BiocViews.html#___Software">BiocViews</a></li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li>Tests</li>
</ol>
<ul>
<li><a href="http://bioconductor.org/developers/how-to/unitTesting-guidelines/">Unit test guidelines</a></li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li><p>Complete and detailed vignette(s) and man pages, with executable examples</p></li>
<li><p>Check time &lt; 5 minutes</p></li>
<li><p>Package size &lt; 5Mb</p></li>
<li><p>All package guidelines can be found <a href="http://bioconductor.org/developers/package-guidelines/">here</a></p></li>
</ol>
<p><strong>IMPORTANT</strong>: A clean build, check, and BiocCheck is not a guaranteed acceptance. The package will still go through a formal review process.</p>
</div>
<div id="submitting-to-bioconductor" class="section level3">
<h3 class="hasAnchor">
<a href="#submitting-to-bioconductor" class="anchor"></a>Submitting to Bioconductor</h3>
<p>Be sure to read the <a href="https://github.com/Bioconductor/Contributions">Contributions Page</a> and when you are ready to submit open a <a href="https://github.com/Bioconductor/Contributions/issues">New Issue</a>. The Title: should be the name of your package. Once the package is approved for building, don’t forget to set up the <a href="https://bioconductor.org/developers/how-to/git/new-package-workflow/">remotes</a>. Some details about the review process can be found <a href="http://bioconductor.org/developers/package-submission/#whattoexpect">here</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kayla Interdonato, Lori Shepherd.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
